vllm/model_executor/layers/quantization/modelopt.py

```diff
@@ -162,3 +167,237 @@ def apply(
                                      input_scale=layer.input_scale,
                                      bias=bias)
+
+
+class ModelOptNvFp4Config(QuantizationConfig):
+    """Config class for ModelOpt FP4."""
+
+    def __init__(self,
+                 is_checkpoint_nvfp4_serialized: bool = False,
+                 kv_cache_quant_algo: str = "",
+                 group_size: int = 16,
+                 exclude_modules: List[str] = None) -> None:
+        self.is_checkpoint_nvfp4_serialized = is_checkpoint_nvfp4_serialized
+        if is_checkpoint_nvfp4_serialized:
+            logger.warning(
+                "Detected ModelOpt NVFP4 checkpoint. Please note that"
+                " the format is experimental and could change in future.")
+
+            self.group_size = group_size
+            self.kv_cache_quant_algo = kv_cache_quant_algo
+            self.exclude_modules = exclude_modules
+
+    @classmethod
+    def get_name(cls) -> str:
+        return "modelopt_nvfp4"
+
+    @classmethod
+    def get_supported_act_dtypes(cls) -> List[torch.dtype]:
+        return [torch.bfloat16, torch.half, torch.float8_e4m3fn]
+
+    @classmethod
+    def get_min_capability(cls) -> int:
+        return 100
+
+    @classmethod
+    def get_config_filenames(cls) -> List[str]:
+        return ["hf_quant_config.json"]
+
+    @classmethod
+    def from_config(cls, config: Dict[str, Any]) -> "ModelOptNvFp4Config":
+        quant_config = cls.get_from_keys(config, ["quantization"])
+        quant_method = quant_config["quant_algo"]
+        if quant_method not in QUANT_ALGOS:
+            raise ValueError(f"ModelOpt currently only supports: {QUANT_ALGOS}"
+                             " quantizations in vLLM. Please check the "
+                             "`hf_quant_config.json` file for your model's "
+                             "quant configuration.")
+        is_checkpoint_nvfp4_serialized = ("NVFP4" in quant_method)
+        kv_cache_quant_algo = quant_config["kv_cache_quant_algo"]
+        group_size = quant_config["group_size"]
+        exclude_modules = quant_config["exclude_modules"]
+        if not (group_size and kv_cache_quant_algo and exclude_modules):
+            raise ValueError("NVFP4 quantization requires group size and "
+                             "kv_cache_quant_algo specified in "
+                             "hf_quant_config.json")
+        return cls(is_checkpoint_nvfp4_serialized, kv_cache_quant_algo,
+                   group_size, exclude_modules)
+
+    def get_quant_method(self, layer: torch.nn.Module,
+                         prefix: str) -> Optional["QuantizeMethodBase"]:
+        from vllm.attention.layer import Attention  # Avoid circular import
+        if isinstance(layer, LinearBase):
+            return ModelOptNvFp4LinearMethod(self)
+        elif isinstance(layer, Attention):
+            return ModelOptFp8KVCacheMethod(self)
+        return None
```

Comment:
It looks like we are missing the check between prefix and exclude_modules to ignore layers, as noted in the model config https://huggingface.co/nvidia/DeepSeek-R1-FP4/blob/761db7ea1b0b750e29fa78ebd2ce449e619809c5/hf_quant_config.json#L10-L15

Reply:
DeepseekR1-FP4 doesn't work with this PR(We are missing support for FP4 Group Gemm). I added back the `exclude_modules` and now load the linear layer classes accordingly. The actual correctness will be checked after adding support for the R1-FP4 model in a future PR.